Post describing the verbatim invalid deduplication bug (#57)
ethereum/solidity-website: src/posts/2023-11-08-verbatim-invalid-deduplication-bug.md

---
layout: post
published: true
title: 'Bug in Block Deduplicator Considering Verbatim Blocks Identical'
date: '2023-11-08'
author: Solidity Team
category: Security Alerts
---

> On October 24, Ori Pomerantz reported a bug regarding the use of `verbatim`

Reviewer: Well, it probably makes sense, but just wanted to note that we almost never do that in those blog posts, not sure why.

rest of the block is then formed by the sequence of assembly opcodes until the
first `JUMP` or other opcode capable of altering the control flow, such as
`RETURN` or `REVERT`.
When the optimizer encounters a tag id `Ta`, it identifies the beginning of a
block of items starting at a certain position `i`. Then, the optimizer checks
if there is a previously seen block of identical assembly items, which starts at
position `j` and is referenced by a tag `Tb`. In case such a block exists, the
optimizer maps `Tb` as a replacement for the tag `Ta`, which means that later
one block will be not referred to by any tag and can then be erased.
If no such block exists, the block referred to by `Ta` gets registered as a
previously seen block and the next one is found and processed.

The bug manifested when checking if a given block `i` is identical to a
previously seen block `j`. The comparison between two assembly blocks is based
on the [operator `<`](https://github.com/ethereum/solidity/blob/v0.8.22/libevmasm/AssemblyItem.h#L136-L146)
of assembly item. The bug only affected `verbatim` items, as their bytecode
contents were incorrectly compared using operator `==`, instead of operator `<`.

> For example, the Yul code snippet showed before, when compiled **without**
> optimizations will include the assembly code fragment shown below:

Reviewer 2: ```suggestion
For example, when the Yul code snippet shown before is compiled **without**
optimizations, it will include the assembly code fragment shown below:
```

small nitpick. does it read better?

```bash
solc --strict-assembly verbatim.yul --asm
```

```
tag_2:
  dup4
  verbatimbytecode_506000
  swap2
  pop
  jump(tag_1)
tag_3:
  0x01
  swap2
  pop
  jump(tag_1)
tag_4:
  dup4
  verbatimbytecode_506002
  swap2
  pop
  jump(tag_1)
```

Notice that the blocks referred to by `tag_2` and `tag_4` are almost identical.

> If such code would be optimized, when the deduplicator looks at the sequence of
> both blocks, due to the bug, the `verbatim` items were considered equal, despite
> their different contents. That happened because, as explained before, the contents
> should be compared using the operator `<`, but for `verbatim` items, that was
> incorrectly implemented by comparing their contents using operator `==`.
> So, when comparing the contents of the `verbatim` items of the example,
> `506000` < `506002` was `false` as was `506002` < `506000`. In this way,
> the contents were concluded to be equal and thus, the optimizer would map `tag_4`
> as a replacement for `tag_2`, which was clearly a mistake. That would cause the
> block referred to by `tag_2` to be erased later. The result of the bug can be
> seen in the output of the optimized code next. Notice that only the block with
> `verbatimbytecode_506002` instruction is left now:

Reviewer 2: Rewritten to suggest some changes in the sentence structure. Feel free to add or ignore.

```suggestion
As a result of the bug, the deduplication step of the optimizer would consider the `verbatim` items as equals despite their different contents. Note that the `<` operator should be used to make the comparison between the contents of the items. The incorrect comparison of the contents occurs due to the usage of the `==` operator instead of the `<` operator to make the comparison.

You can see that in the example, the comparisons of the `verbatim` items `506000` < `506002` and `506002` < `506000` were false and hence, the contents were concluded to be equal. The optimizer would then _incorrectly_ map `tag_4`
as a replacement for `tag_2` which would further cause the
block referred to by `tag_2` to be erased later.

The result of the bug can be
seen in the output of the optimized code below:

```

```bash
solc --strict-assembly verbatim.yul --asm --optimize
```

```
  0x00
  dup1
  sload
  dup1
  iszero
  tag_5
  jumpi
  dup1
  0x01
  eq
  tag_3
  jumpi
  dup1
  0x02
  eq
  tag_5
  jumpi
  0x03
  eq
  tag_7
  jumpi
  0x00
  sstore
  stop
tag_7:
  pop
  0x03
  0x00
  sstore
  stop
tag_5:
  pop
  pop
  0xaabbccddeeff
  verbatimbytecode_506002
  0x00
  sstore
  stop
tag_3:
  pop
  pop
  0x01
  0x00
  sstore
  stop

> ```

Reviewer 2: Maybe add the following sentence here instead?

> Notice that only the block with
`verbatimbytecode_506002` instruction remains in the output shown above as a result of the bug.

---
layout: post
published: true
title: 'Bug in Deduplication of Verbatim Blocks'
date: '2023-11-08'
author: Solidity Team
category: Security Alerts
---

On October 24, Ori Pomerantz reported a bug affecting the use of `verbatim`
builtin in Yul code. After investigating, the team was able to confirm the problem
and locate its origin.
The bug existed in the Block Deduplicator optimizer step, wherein equivalent
assembly blocks are identified and merged.
`verbatim` assembly items surrounded by identical opcodes were incorrectly considered identical
and unified.

The bug existed since version `0.8.5`, which introduced `verbatim`, and only affected pure Yul
compilation with optimization enabled.
Solidity code or Yul used in inline assembly blocks would not trigger it.
To mitigate it, it is sufficient to turn off the `settings.optimizer.details.deduplicate` flag in the
in [Standard JSON input](https://docs.soliditylang.org/en/v0.8.23/using-the-compiler.html#input-description).

> Although the use of `verbatim` is not uncommon, the conditions which trigger

Reviewer 3: Yeah? I thought we concluded that nobody actually uses `verbatim` at all, didn't we :-)?

Author: True :sweat_smile: 
I will rewrite saying that the use of `verbatim` **is** uncommon and the conditions which trigger it are very specific.

the bug (two or more `verbatim` items surrounded by identical opcodes) are very specific.
Also, to the extent of the investigations made by the team, there is no evidence
that this could be used as an exploit or attack vector.
Considering that, the team assigned the bug a severity level of `low`.

## Which Contracts Are Affected?

> The precise conditions which triggered the bug are as follows:

Reviewer 3: The list below doesn't have "precise conditions", but "necessary conditions" - if we state those as precise decisions, it makes it sound like *any* use of verbatim is affected, while in fact it's still highly unlikely that you hit it if the conditions here are true (you need very precise patterns to achieve fully identical blocks on the evm assembly level - identical there actually means identical *including* outgoing jumps - that can almost only happen in switches and at the very end of functions to begin with)

1. Compilation of pure Yul code.
1. The use of the `verbatim` builtin.
1. Block Deduplicator optimizer step being in use.

Note that Block Deduplicator is enabled by default when the optimizer is enabled.
In Standard JSON input it can be explicitly disabled using the `deduplicate` setting.

> On the command line, it can only be disabled along with the whole opcode-based optimizer.

Reviewer 3: We should not talk about disabling it, that's not a sane thing to do for anyone. Mentioning that it's enabled by default is fine, but let's drop the part about disabling it.

> If your project does not include contracts written purely in Yul,

Reviewer 3: ```suggestion
If your project does not include contracts written purely in Yul or does not use ``verbatim``,
```

which operates on assembly code. It splits the sequence of instructions into
blocks usually separated by `JUMPS` and `JUMPDEST`s. The blocks are analyzed
and a series of optimization steps are performed.
During deduplication step, the optimizer maps which tags can be replaced by one
another, because they refer to blocks of identical sequences of opcodes. The blocks
are considered identical if both have the same opcodes in the same order.

The block deduplicator identifies the start of a block upon finding a tag. The
rest of the block is then formed by the sequence of assembly opcodes until the
first `JUMP` or other opcode capable of altering the control flow, such as
`RETURN` or `REVERT`.
When the optimizer encounters a tag id `Ta`, it identifies the beginning of a
block of items starting at a certain position `i`. Then, the optimizer checks
if there is a previously seen block of identical assembly items, which starts at
position `j` and is referenced by a tag `Tb`. In case such a block exists, the
optimizer maps `Tb` as a replacement for the tag `Ta`, which means that later

> one block will be not referred to by any tag and can then be erased.

Author: ```suggestion
one block will not be referred to by any tag and can then be erased.
```
I guess this is the correct phrasing.

If no such block exists, the block referred to by `Ta` gets registered as a
previously seen block and the next one is found and processed.

The bug manifested when checking if a given block `i` is identical to a
previously seen block `j`. The comparison between two assembly blocks is based
on the [operator `<`](https://github.com/ethereum/solidity/blob/v0.8.22/libevmasm/AssemblyItem.h#L136-L146)
of assembly item. The bug only affected `verbatim` items, as their bytecode

> contents were incorrectly compared using operator `==`, instead of operator `<`.

Reviewer 3: If we want, we can describe the effect instead, which is that for the block deduplicator all verbatim items are equal, that's enough.

> 

Reviewer 3: I'm generally not happy about this description.
This is a very convoluted way of saying it compares two blocks by comparing all assembly items in them in sequence. There is no benefit in following the implementation by tracing indices here, that only makes it harder to understand what's happening.

For example, the Yul code snippet showed before, when compiled **without**
optimizations will include the assembly code fragment shown below:

```bash
solc --strict-assembly verbatim.yul --asm
```

```
tag_2:
  dup4
  verbatimbytecode_506000
  swap2
  pop
  jump(tag_1)
tag_3:
  0x01
  swap2
  pop
  jump(tag_1)
tag_4:
  dup4
  verbatimbytecode_506002
  swap2
  pop
  jump(tag_1)
```

Notice that the blocks referred to by `tag_2` and `tag_4` are almost identical.

> If such code would be optimized, when the deduplicator looks at the sequence of

Author: ```suggestion
If such code would be optimized, when the deduplicator looked at the sequence of
```

both blocks, due to the bug, the `verbatim` items were considered equal, despite
their different contents. That happened because, as explained before, the contents
should be compared using the operator `<`, but for `verbatim` items, that was

> incorrectly implemented by comparing their contents using operator `==`.

Reviewer 3: Again, that's too much technical details about the comparison here.
